lib/tpm2_auth_util.c

```diff
@@ -110,48 +87,77 @@ static bool handle_str(const char *password, TPMS_AUTH_COMMAND *auth) {
     return true;
 }
 
-bool tpm2_auth_util_from_optarg(TSS2_SYS_CONTEXT *sys_ctx, const char *password,
-    TPMS_AUTH_COMMAND *auth, tpm2_session **session) {
+static bool handle_password(const char *password, TPMS_AUTH_COMMAND *auth) {
 
-    char *password_string_ptr;
-    char *password_string = strdup(password);
-    if (!password_string) {
-        LOG_ERR("OOM");
-        return false;
+    bool is_hex = !strncmp(password, HEX_PREFIX, HEX_PREFIX_LEN);
+    if (is_hex) {
+        return handle_hex_password(password, auth);
     }
-    password_string_ptr = password_string;
 
-    bool result;
-    bool is_hex;
-    bool is_session;
+    /* must be string, handle it */
+    return handle_str_password(password, auth);
+}
 
-    const char *password_token;
+static bool handle_session(TSS2_SYS_CONTEXT *sys_ctx, const char *path, TPMS_AUTH_COMMAND *auth,
+        tpm2_session **session) {
 
-    while( (password_token = strsep(&password_string,"+")) != NULL ) {
-        is_hex = !strncmp(password_token, HEX_PREFIX, HEX_PREFIX_LEN);
-        is_session = !strncmp(password_token, SESSION_PREFIX, SESSION_PREFIX_LEN);
-        if (is_session && !session) {
-            LOG_ERR("Tool does not support sessions for this auth value");
-            return false;
-        }
+    /* if it is session, then skip the prefix */
+    path += SESSION_PREFIX_LEN;
 
-        if (is_hex) {
-            result = handle_hex(password_token, auth);
-        } else if (is_session) {
-            result = handle_session(sys_ctx, password_token, auth, session);
-        } else {
-            /* must be string, handle it */
-            result = handle_str(password_token, auth);
-        }
+    /* Make a local copy for manipulation */
+    char tmp[PATH_MAX];
+    size_t len = snprintf(tmp, sizeof(tmp), "%s", path);
+    if (len >= sizeof(tmp)) {
+        LOG_ERR("Path truncated");
+        return false;
+    }
 
-        if (result) {
-            continue;
-        } else {
-            free((char*)password_string_ptr);
-            return false;
+    /*
+     * Sessions can have an associated password/auth value denoted adter
+     * a + sign, ie session.ctx+foo, deal with it by
```

Comment:
is the order restricted here. As in can one do hex:password+session:session.ctx?

Reply:
No that wouldn't work. the flow is session->password as in the session handler we check if their is a +, and if it contains an entry, call the handle_password routine which ONLY handles passwords.

Reply:
Cool. Can you please add this info to https://github.com/tpm2-software/tpm2-tools/blob/master/man/common/authorizations.md?